spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/firestore/GcpFirestoreEmulatorAutoConfiguration.java

```diff
@@ -0,0 +1,152 @@
+import org.springframework.cloud.gcp.autoconfigure.firestore.GcpFirestoreAutoConfiguration.FirestoreReactiveAutoConfiguration;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.cloud.gcp.data.firestore.FirestoreTemplate;
+import org.springframework.cloud.gcp.data.firestore.mapping.FirestoreClassMapper;
+import org.springframework.cloud.gcp.data.firestore.mapping.FirestoreMappingContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides autoconfiguration to use the Firestore emulator if enabled.
+ *
+ * @since 1.2.3
+ * @author Daniel Zou
+ */
+@Configuration
+@ConditionalOnProperty("spring.cloud.gcp.firestore.host-port")
+@AutoConfigureBefore({GcpFirestoreAutoConfiguration.class})
+@EnableConfigurationProperties(GcpFirestoreProperties.class)
+public class GcpFirestoreEmulatorAutoConfiguration {
+
+	private static final String ROOT_PATH_FORMAT = "projects/%s/databases/(default)/documents";
+
+	private final String projectId;
+
+	private final String hostPort;
+
+	private final String firestoreRootPath;
+
+	GcpFirestoreEmulatorAutoConfiguration(
+			GcpFirestoreProperties properties,
+			GcpProjectIdProvider projectIdProvider) throws IOException {
+		this.projectId = projectIdProvider.getProjectId();
+		this.hostPort = properties.getHostPort();
+		this.firestoreRootPath = String.format(ROOT_PATH_FORMAT, this.projectId);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public FirestoreOptions firestoreOptions() {
+		return FirestoreOptions.newBuilder()
+				.setCredentials(emulatorCredentials())
+				.setProjectId(this.projectId)
+				.setChannelProvider(
+						InstantiatingGrpcChannelProvider.newBuilder()
+								.setEndpoint(this.hostPort)
+								.setChannelConfigurator(input -> input.usePlaintext())
+								.build())
+				.build();
+	}
+
+	private static Credentials emulatorCredentials() {
+		final Map<String, List<String>> headerMap = new HashMap<>();
+		headerMap.put("Authorization", Collections.singletonList("Bearer owner"));
+		headerMap.put("google-cloud-resource-prefix", Collections.singletonList("projects/my-project/databases/(default)"));
+
+		return new Credentials() {
+			@Override
+			public String getAuthenticationType() {
+				return null;
+			}
+
+			@Override
+			public Map<String, List<String>> getRequestMetadata(URI uri) {
+				return headerMap;
+			}
+
+			@Override
+			public boolean hasRequestMetadata() {
+				return true;
+			}
+
+			@Override
+			public boolean hasRequestMetadataOnly() {
+				return true;
+			}
+
+			@Override
+			public void refresh() {
+				// no-op
+			}
+		};
+	}
+
+	/**
+	 * Reactive Firestore autoconfiguration to enable emulator use.
+	 */
+	@ConditionalOnClass({ FirestoreGrpc.FirestoreStub.class, Flux.class })
+	@AutoConfigureBefore({ FirestoreReactiveAutoConfiguration.class })
+	class ReactiveFirestoreEmulatorAutoConfiguration {
```

Comment:
static?

Reply:
Right now I depend on the `hostPort` instance variable of the parent so I couldn't make static unfortunately.